tasks/get_sql_logins.ps1

```diff
@@ -0,0 +1,194 @@
+
+}
+
+$error = @{
+    _error = @{
+        msg = ''
+        kind = 'puppetlabs.task/task-error'
+        details = @{
+            detailedInfo = ''
+            exitcode = 1
+        }
+    }
+}
+
+$return = @{}
+
+#Get SQL Instances
+if($MyInvocation.BoundParameters.ContainsKey('instance_name')) {
+    $SQLInstances = $instance_name
+} else {
+    try{
+        $SQLInstances = Get-SQLInstances
+    } catch {
+        $error._error.msg = 'Cannot open registry key HKLM:\SOFTWARE\Microsoft\Microsoft SQL Server'
+        $error._error.details.detailedInfo = $_
+        return $error | ConvertTo-JSON
+    }
+}
+
+# Unfiltered Logins from all instances.
+$rawLogins = @()
+
+foreach($instance in $SQLInstances){
+    if($instance -eq 'MSSQLSERVER'){$instance = '.'}
+    try {
+        $sqlServer = Get-ServerObject -instance $instance
+    } catch {
+        $error._error.msg = "Cannot connect to SQL Instance: $instance"
+        $error._error.details.detailedInfo = $_
+        return $error | ConvertTo-JSON
+    }
+
+    foreach($item in $sqlServer.logins){
+        # The login object doesn't return information about which instance it
+        # came from. This could be a problem on a machine running more than
+        # once instance. We'll add a property here to make sure we don't lose
+        # track of this information.
+
+        Add-Member -InputObject $item -MemberType NoteProperty -Name 'InstanceName' -Value $instance
+
+        $rawLogins += $item
+    }
+}
+
+# Filtered logins based on the filter passed into the $LoginName parameter
+$logins = @()
+
+foreach ($login in $rawLogins) {
+    if($MyInvocation.BoundParameters.ContainsKey('LoginName')) {
+        $logins += Select-LoginName -login $login -namesToMatch $LoginName -ExactMatch:$ExactMatch
+    } else {
+        $logins += $login
+    }
+}
+
+if($Detailed) {
+
+    # The SID property of the Login object contains an array of integers. This
+    # is not how SQL Server represents the login int he sys.server_principals
+    # table. The array of integers needs to be converted one by one into their
+    # Hex representation, and then joined together. This will match the value
+    # in the server_principals table and allow someone executing this script
+    # to correlate users properly.
+    $sidFunction = {
+        $finalSid = '0x'
+        foreach($segment in $_.sid){
+            $finalSid += ("{0:x2}" -f $segment).ToUpper()
+        }
+        $finalSid
+    }
+
+    # The SID column of sys.server_principals stores the binary representation
+    # of an account SID. This is not a SID that can easily be correlated to an
+    # Active Directory style SID that you can find in most other tools. This
+    # function converts SQLServer's binary sid into a SID string that people
+    # are more familiar with.
+    $winSidFunction = {
+        (New-Object System.Security.Principal.SecurityIdentifier($_.sid, 0)).toString()
```

Comment:
I think this may be the wrong approach. I think this may only work if that sid I'm trying to convert is a current, SID. Part of the point here is to detect cases where a SID stored in SQL Server is no longer the correct SID for a given AD User. I may have to find the conversion algorithm somewhere to convert the stored binary SID into the string representing what the AD SID, SHOULD be. I'll have to dig deeper into the documentation for this type to figure out if this valid or not.